src/Eventso.Subscription.Kafka/DeadLetter/PoisonEventQueue.cs

```diff
@@ -0,0 +1,166 @@
+                $"Poison event: {@event.TopicPartitionOffset}. " +
+                $"Error: {failureReason}",
+                null);
+        
+        await poisonEventStore.AddEvent(groupId, @event, failureTimestamp, failureReason, token);
+        
+        var keys = await GetTopicPartitionKeys(@event.TopicPartitionOffset.TopicPartition, token);
+        keys.Add(key);
+    }
+
+    public async IAsyncEnumerable<ConsumeResult<byte[], byte[]>> Peek([EnumeratorCancellation] CancellationToken token)
+    {
+        bool needToProcess = true;
+
+        while (needToProcess)
+        {
+            needToProcess = false;
+            var topicPartitions = _partitionPoisonKeys.Keys.ToArray();
+            foreach (var topicPartition in topicPartitions)
+            {
+                try
+                {
+                    var keys = await GetTopicPartitionKeys(topicPartition, token);
+                    if (keys.Count == 0)
+                        continue;
+                }
+                catch (Exception)
+                {
+                    continue;
+                }
+
+                var eventForRetrying = await retryScheduler.GetNextRetryTarget(groupId, topicPartition, token);
+                if (eventForRetrying == null)
+                    continue;
+
+                needToProcess = true;
+                yield return eventForRetrying;
+            }
+        }
+    }
+
+    public async Task Dequeue(ConsumeResult<byte[], byte[]> @event, CancellationToken token)
+    {
+        var key = DeserializeKey(@event.Topic, @event.Message.Headers, @event.Message.Key);
+        logger.Dequeue(
+            groupId,
+            @event.TopicPartitionOffset.Topic,
+            @event.TopicPartitionOffset.Partition,
+            @event.TopicPartitionOffset.Offset,
+            key);
+        await poisonEventStore.RemoveEvent(groupId, @event.TopicPartitionOffset, token);
+        if (await poisonEventStore.IsKeyPoisoned(groupId, @event.TopicPartitionOffset.Topic, @event.Message.Key, token))
+            return;
+
+        var keys = await GetTopicPartitionKeys(@event.TopicPartitionOffset.TopicPartition, token);
+        keys.Remove(key);
+    }
+
+    private static Guid DeserializeKey(string topic, Headers headers, byte[]? keyBytes)
+    {
+        return KeyGuidDeserializer.Instance.Deserialize(
+            keyBytes,
+            keyBytes == null,
+            new SerializationContext(MessageComponentType.Key, topic, headers));
+    }
+
+    private async ValueTask<HashSet<Guid>> GetTopicPartitionKeys(TopicPartition topicPartition, CancellationToken token)
+    {
+        if (!_partitionPoisonKeys.TryGetValue(topicPartition, out var keysTask))
+            throw new Exception($"{topicPartition} disabled");
+
+        if (!keysTask.Task.IsCompleted)
+        {
+            using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                keysTask.CancellationTokenSource.Token,
```

Comment:
doesn't make much sense to include this token, bc keysTask.Task is actually cancelled by it

Reply:
all in all it's just `return keysTask.Task.WaitAsync(token);`

Reply:
++